spring-boot-project/spring-boot/src/main/java/org/springframework/boot/webservices/client/WebServiceTemplateBuilder.java

```diff
@@ -0,0 +1,1106 @@
+			else if (type == Timeout.READ) {
+				customizer.setReadTimeout(target, timeout);
+			}
+		}
+
+		interface TimeoutCustomizer<T> {
+
+			void setReadTimeout(T source, int timeout);
+
+			void setConnectionTimeout(T source, int timeout);
+
+		}
+
+		enum Timeout {
+
+			READ, CONNECTION
+
+		}
+
+		private static final class HttpComponentsTimeoutCustomizer
+				implements TimeoutCustomizer<HttpComponentsMessageSender> {
+
+			@Override
+			public void setReadTimeout(HttpComponentsMessageSender source, int timeout) {
+				source.setReadTimeout(timeout);
+			}
+
+			@Override
+			public void setConnectionTimeout(HttpComponentsMessageSender source,
+					int timeout) {
+				source.setConnectionTimeout(timeout);
+			}
+
+		}
+
+		private static final class HttpUrlConnectionTimeoutCustomizer
+				implements TimeoutCustomizer<HttpUrlConnectionMessageSender> {
+
+			@Override
+			public void setReadTimeout(HttpUrlConnectionMessageSender source,
+					int timeout) {
+				source.setReadTimeout(Duration.ofMillis(timeout));
+			}
+
+			@Override
+			public void setConnectionTimeout(HttpUrlConnectionMessageSender source,
+					int timeout) {
+				source.setConnectionTimeout(Duration.ofMillis(timeout));
+			}
+
+		}
+
+		private static final class ClientHttpRequestTimeoutCustomizer
+				implements TimeoutCustomizer<ClientHttpRequestMessageSender> {
+
+			private static final Map<String, Class<? extends TimeoutCustomizer<? extends ClientHttpRequestFactory>>> CUSTOMIZERS;
+
+			static {
+				Map<String, Class<? extends TimeoutCustomizer<? extends ClientHttpRequestFactory>>> candidates = new LinkedHashMap<>();
+				candidates.put(
+						"org.springframework.http.client.HttpComponentsClientHttpRequestFactory",
+						HttpComponentsClientHttpRequestFactoryTimeoutCustomizer.class);
+				candidates.put(
+						"org.springframework.http.client.OkHttp3ClientHttpRequestFactory",
+						OkHttp3ClientHttpRequestFactoryTimeoutCustomizer.class);
+				candidates.put(
+						"org.springframework.http.client.SimpleClientHttpRequestFactory",
+						SimpleClientHttpRequestFactoryTimeoutCustomizer.class);
+				CUSTOMIZERS = Collections.unmodifiableMap(candidates);
+			}
+
+			@Override
+			public void setReadTimeout(ClientHttpRequestMessageSender source,
+					int timeout) {
+				ClassLoader classLoader = getClass().getClassLoader();
+				customize(CUSTOMIZERS, getRequestFactory(source), Timeout.READ, timeout,
+						classLoader);
+			}
+
+			@Override
+			public void setConnectionTimeout(ClientHttpRequestMessageSender source,
+					int timeout) {
+				ClassLoader classLoader = getClass().getClassLoader();
+				customize(CUSTOMIZERS, getRequestFactory(source), Timeout.CONNECTION,
+						timeout, classLoader);
+			}
+
+			private ClientHttpRequestFactory getRequestFactory(
+					ClientHttpRequestMessageSender source) {
+				ClientHttpRequestFactory requestFactory = source.getRequestFactory();
+				if (!(requestFactory instanceof AbstractClientHttpRequestFactoryWrapper)) {
```

Comment:
I haven't looked in detail but I am wondering why you need to resort to reflection for a builder pattern that should "just" do what the regular template can do. Perhaps this feature should be removed in benefit from something more advanced? I don't like the idea to use reflection in production code.